cumulusci/tasks/connectedapp.py

```diff
@@ -0,0 +1,174 @@
+import json
+import re
+import shutil
+import os
+import tempfile
+from cumulusci.core.config import ServiceConfig
+from cumulusci.core.exceptions import TaskOptionsError, ServiceNotConfigured
+from cumulusci.core.utils import process_bool_arg
+from cumulusci.tasks.sfdx import SFDXBaseTask, SFDX_CLI
+from cumulusci.utils import random_alphanumeric_underscore
+
+CONNECTED_APP = """<?xml version="1.0" encoding="UTF-8"?>
+<ConnectedApp xmlns="http://soap.sforce.com/2006/04/metadata">
+    <contactEmail>{email}</contactEmail>
+    <label>{label}</label>
+    <oauthConfig>
+        <callbackUrl>http://localhost:8080/callback</callbackUrl>
+        <consumerKey>{client_id}</consumerKey>
+        <consumerSecret>{client_secret}</consumerSecret>
+        <scopes>Web</scopes>
+        <scopes>Full</scopes>
+        <scopes>RefreshToken</scopes>
+    </oauthConfig>
+</ConnectedApp>"""
+
+PACKAGE_XML = """<?xml version="1.0" encoding="UTF-8"?>
+<Package xmlns="http://soap.sforce.com/2006/04/metadata">
+    <types>
+        <members>*</members>
+        <name>ConnectedApp</name>
+    </types>
+    <version>44.0</version>
+</Package>"""
+
+
+class CreateConnectedApp(SFDXBaseTask):
+    task_options = {
+        "label": {
+            "description": "The label for the connected app.  Must contain only alphanumeric and underscores",
+            "required": True,
+        },
+        "email": {
+            "description": "The email address to associate with the connected app.  Defaults to email address from the github service if configured.",
+            "required": True,
+        },
+        "username": {
+            "description": "Create the connected app in a different org.  Defaults to the defaultdevhubusername configured in sfdx.",
+            "required": False,
+        },
+        "connect": {
+            "description": "If True, the created connected app will be stored as the CumulusCI connected_app service in the keychain.",
+            "required": False,
+        },
+        "overwrite": {
+            "description": "If True, any existing connected_app service in the CumulusCI keychain will be overwritten.  Has no effect if the connect option is False.",
+            "required": False,
+        },
+    }
+
+    def _init_options(self, kwargs):
+        self.client_id = None
+        self.client_secret = None
+        kwargs["command"] = "force:mdapi:deploy --wait 5"
+        super(CreateConnectedApp, self)._init_options(kwargs)
+
+        # Validate label
+        if not re.match(r"^\w+$", self.options["label"]):
+            raise TaskOptionsError(
+                "label value must contain only alphanumeric or underscore characters"
+            )
+
+        # Default email to the github service's email if configured
```

Comment:
the SAL team uses this little "hack" of adding email to the porject_conf for their scripts: https://github.com/SalesforceFoundation/sal/blob/master/dev_config/readme.md#use-custom-email-in-dev-data.

i like getting the github user email, personally!